x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -7,63 +7,90 @@
   Logger,
 } from 'src/core/server';
 import { ConfigType } from '../index';
 
-interface IEnterpriseSearchRequestParams<ResponseBody> {
+interface IConstructorDependencies {
   config: ConfigType;
   log: Logger;
+}
+interface IRequestParams<ResponseBody> {
   path: string;
+  params?: string;
   hasValidData?: (body?: ResponseBody) => boolean;
 }
+export interface IEnterpriseSearchRequestHandler {
+  createRequest(requestParams?: object): RequestHandler<unknown, Readonly<{}>, unknown>;
+}
 
 /**
- * This helper function creates a single standard DRY way of handling
+ * This helper lib creates a single standard DRY way of handling
  * Enterprise Search API requests.
  *
  * This handler assumes that it will essentially just proxy the
  * Enterprise Search API request, so the request body and request
  * parameters are simply passed through.
  */
-export function createEnterpriseSearchRequestHandler<ResponseBody>({
-  config,
-  log,
-  path,
-  hasValidData = () => true,
-}: IEnterpriseSearchRequestParams<ResponseBody>) {
-  return async (
-    _context: RequestHandlerContext,
-    request: KibanaRequest<unknown, Readonly<{}>, unknown>,
-    response: KibanaResponseFactory
-  ) => {
-    try {
-      const enterpriseSearchUrl = config.host as string;
-      const params = request.query ? `?${querystring.stringify(request.query)}` : '';
-      const url = `${encodeURI(enterpriseSearchUrl)}${path}${params}`;
+export class EnterpriseSearchRequestHandler {
+  private enterpriseSearchUrl: string;
+  private log: Logger;
 
-      const apiResponse = await fetch(url, {
-        headers: { Authorization: request.headers.authorization as string },
-      });
+  constructor({ config, log }: IConstructorDependencies) {
+    this.log = log;
+    this.enterpriseSearchUrl = config.host as string;
+  }
 
-      const body = await apiResponse.json();
+  createRequest<ResponseBody>({
+    path,
+    params,
+    hasValidData = () => true,
+  }: IRequestParams<ResponseBody>) {
+    return async (
+      _context: RequestHandlerContext,
+      request: KibanaRequest<unknown, Readonly<{}>, unknown>,
+      response: KibanaResponseFactory
+    ) => {
+      try {
+        // Set up API URL
+        params = params ?? (request.query ? `?${querystring.stringify(request.query)}` : '');
+        const url = encodeURI(this.enterpriseSearchUrl + path + params);
 
-      if (hasValidData(body)) {
-        return response.ok({ body });
-      } else {
-        throw new Error(`Invalid data received: ${JSON.stringify(body)}`);
-      }
-    } catch (e) {
-      log.error(`Cannot connect to Enterprise Search: ${e.toString()}`);
-      if (e instanceof Error) log.debug(e.stack as string);
+        // Set up API options
+        const { method } = request.route;
+        const headers = { Authorization: request.headers.authorization as string };
+        const body = Object.keys(request.body as object).length
+          ? JSON.stringify(request.body)
+          : undefined;
```

Comment:
Follow up to the above comment:

Kibana's `request.body` sends an empty obj, while node-fetch's expects an `undefined`, hence the error.

We might also want to consider checking for `method === 'GET'`/`HEAD` and not allowing a body for GET requests (see below repeated screenshot), but that's maybe overkill? What do you think?

<img width="727" alt="Screen Shot 2020-08-26 at 4 53 40 PM" src="https://user-images.githubusercontent.com/549407/91466072-79139200-e843-11ea-8708-c8a9e6b96ef8.png">